pkg/controller/flink/client/error_handler.go

```diff
@@ -0,0 +1,160 @@
+package client
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// appError codes
+const (
+	GlobalFailure      = "FAILED"
+	JSONUnmarshalError = "JSONUNMARSHALERROR"
+	defaultRetries     = 20
+	noRetries          = 0
+)
+
+type FlinkMethod string
+
+const (
+	CancelJobWithSavepoint FlinkMethod = "CancelJobWithSavepoint"
+	ForceCancelJob         FlinkMethod = "ForceCancelJob"
+	SubmitJob              FlinkMethod = "SubmitJob"
+	CheckSavepointStatus   FlinkMethod = "CheckSavepointStatus"
+	GetJobs                FlinkMethod = "GetJobs"
+	GetClusterOverview     FlinkMethod = "GetClusterOverview"
+	GetLatestCheckpoint    FlinkMethod = "GetLatestCheckpoint"
+	GetJobConfig           FlinkMethod = "GetJobConfig"
+	GetTaskManagers        FlinkMethod = "GetTaskManagers"
+	GetCheckpointCounts    FlinkMethod = "GetCheckpointCounts"
+	GetJobOverview         FlinkMethod = "GetJobOverview"
+)
+
+// FlinkApplicationError implements the error interface to make error handling more structured
+type FlinkApplicationError struct {
+	AppError            string       `json:"appError,omitempty"`
+	Method              FlinkMethod  `json:"method,omitempty"`
+	ErrorCode           string       `json:"errorCode,omitempty"`
+	IsRetryable         bool         `json:"isRetryable,omitempty"`
+	IsFailFast          bool         `json:"isFailFast,omitempty"`
+	MaxRetries          int32        `json:"maxRetries,omitempty"`
+	LastErrorUpdateTime *metav1.Time `json:"startedAt,omitempty"`
+}
+
+func (f *FlinkApplicationError) Error() string {
+	return f.AppError
+}
+
+func (f *FlinkApplicationError) DeepCopyInto(out *FlinkApplicationError) {
+	*out = *f
+	if f.LastErrorUpdateTime != nil {
+		f, out := &f.LastErrorUpdateTime, &out.LastErrorUpdateTime
+		*out = (*f).DeepCopy()
+	}
+}
+
+func (f *FlinkApplicationError) DeepCopy() *FlinkApplicationError {
+	if f == nil {
+		return nil
+	}
+	out := new(FlinkApplicationError)
+	f.DeepCopyInto(out)
+	return out
+}
+
+func GetError(err error, method FlinkMethod, errorCode string, isRetryable bool, isFailFast bool, maxRetries int32, message ...string) error {
```

Comment:
I'd prefer if we can avoid having these two boolean arguments here. It's very hard to tell what they mean from reading function calls, and easy to mix them up.

Maybe instead have methods like GetRetryableError / GetFastFailError?

Reply:
Fair point. Refactored to make these two separate methods.